### PR TITLE
Module2

### PR DESCRIPTION
**Pull request by letucennik**

### CODE REVIEW

File: repository/pom.xml [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+                <plugin>
+                    <artifactId>maven-war-plugin</artifactId>
+                    <version>2.3</version>
+                    <configuration>
+                        <failOnMissingWebXml>false</failOnMissingWebXml>
+                    </configuration>
+                </plugin>
+            </plugins>
+        </pluginManagement>
+    </build>
```
Comment (SenPavel) [inline repository/pom.xml]: Is the "war" plugin needed in the repository?

File: repository/src/test/java/com/epam/esm/repository/impl/CertificateTagRepositoryImplTest.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import org.springframework.test.context.ContextConfiguration;
+import org.springframework.test.context.junit.jupiter.SpringExtension;
+
+import java.util.Collections;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+@ExtendWith(SpringExtension.class)
+@Component
```
Comment (SenPavel) [inline repository/src/test/java/com/epam/esm/repository/impl/CertificateTagRepositoryImplTest.java]: The Component annotation is not needed here

File: repository/src/test/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImplTest.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+import org.springframework.test.context.junit.jupiter.SpringExtension;
+
+import java.math.BigDecimal;
+import java.time.LocalDateTime;
+import java.util.*;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+@ExtendWith(SpringExtension.class)
+@Component
```
Comment (SenPavel) [inline repository/src/test/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImplTest.java]: the same

File: repository/src/test/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImplTest.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+@ContextConfiguration(classes = {TestJdbcConfig.class})
+@ComponentScan("com.epam.esm.entity.mapper")
```
Comment (SenPavel) [inline repository/src/test/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImplTest.java]: If component scan is needed, move it to TestJdbcConfig

File: repository/src/test/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImplTest.java [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    @Test
+    void testShouldCreate() {
+        Long id = giftCertificateRepository.create(certificateToCreate);
+        assertNotNull(id);
+    }
+
+    @Test
+    void testShouldFindById() {
+        Optional<GiftCertificate> giftCertificate = giftCertificateRepository.read(1);
+        assertTrue(giftCertificate.isPresent() && giftCertificate.get().getId() == firstCertificate.getId());
```
Comment (SenPavel) [inline repository/src/test/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImplTest.java]: Please divide this assert into two asserts, assertTrue for isPresent and assertEquals for giftCertificate.get().getId() == firstCertificate.getId()

File: repository/src/test/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImplTest.java [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+        assertFalse(giftCertificate.isPresent());
+    }
+
+    @Test
+    void testShouldUpdateById() {
+        Map<String, Object> updateInfo = new HashMap<>();
+        updateInfo.put("name", "certificate new name");
+        giftCertificateRepository.update(firstCertificate.getId(), updateInfo);
+        Optional<GiftCertificate> certificate = giftCertificateRepository.read(firstCertificate.getId());
+        assertTrue(certificate.isPresent() && certificate.get().getName().equals("certificate new name"));
```
Comment (SenPavel) [inline repository/src/test/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImplTest.java]: the same

File: repository/src/test/java/com/epam/esm/repository/impl/TagRepositoryImplTest.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+import org.springframework.stereotype.Component;
+import org.springframework.test.context.ContextConfiguration;
+import org.springframework.test.context.junit.jupiter.SpringExtension;
+
+import java.util.Optional;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+@ExtendWith(SpringExtension.class)
+@Component
```
Comment (SenPavel) [inline repository/src/test/java/com/epam/esm/repository/impl/TagRepositoryImplTest.java]: the same

File: repository/src/test/java/com/epam/esm/repository/impl/TagRepositoryImplTest.java [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+    @Test
+    void testShouldCreate() {
+        Long id = tagRepository.create(tagToCreate);
+        assertNotNull(id);
+    }
+
+    @Test
+    void testShouldFindById() {
+        Optional<Tag> tag = tagRepository.read(firstTag.getId());
+        assertTrue(tag.isPresent() && tag.get().equals(firstTag));
```
Comment (SenPavel) [inline repository/src/test/java/com/epam/esm/repository/impl/TagRepositoryImplTest.java]: the same for asserts

File: service/pom.xml [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+                <plugin>
+                    <artifactId>maven-war-plugin</artifactId>
+                    <version>2.3</version>
+                    <configuration>
+                        <failOnMissingWebXml>false</failOnMissingWebXml>
+                    </configuration>
+                </plugin>
+            </plugins>
+        </pluginManagement>
+    </build>
```
Comment (SenPavel) [inline service/pom.xml]: Is it needed in the service?

File: web/pom.xml [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+                </plugin>
+                <!-- site lifecycle, see https://maven.apache.org/ref/current/maven-core/lifecycles.html#site_Lifecycle -->
+                <plugin>
+                    <artifactId>maven-site-plugin</artifactId>
+                    <version>3.7.1</version>
+                </plugin>
+                <plugin>
+                    <artifactId>maven-project-info-reports-plugin</artifactId>
+                    <version>3.0.0</version>
+                </plugin>
```
Comment (SenPavel) [inline web/pom.xml]: Are these plugins needed?

**Review by SenPavel — CHANGES_REQUESTED** [review]

File: repository/src/main/java/com/epam/esm/config/DataSourceConfig.java [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+        return config;
+    }
+
+    @Bean
+    public KeyHolder keyHolder() {
+        return new GeneratedKeyHolder();
+    }
+
+    @Bean
+    @Autowired
```
Comment (maximelyashevich) [inline repository/src/main/java/com/epam/esm/config/DataSourceConfig.java]: Please check if this annotation is really needed here.

File: repository/src/main/java/com/epam/esm/repository/impl/CertificateTagRepositoryImpl.java [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+
+    private static final String CREATE_CERTIFICATE_TAG_REFERENCE = "INSERT INTO m2m_certificates_tags (gift_certificate_id, tag_id) VALUES (?, ?)";
+    private static final String FIND_TAGS_ID_BY_CERTIFICATE_ID = "SELECT tag_id FROM m2m_certificates_tags WHERE gift_certificate_id=?";
+
+    private final JdbcTemplate jdbcTemplate;
+
+    @Autowired
+    public CertificateTagRepositoryImpl(JdbcTemplate jdbcTemplate) {
+        this.jdbcTemplate = jdbcTemplate;
+    }
```
Comment (maximelyashevich) [inline repository/src/main/java/com/epam/esm/repository/impl/CertificateTagRepositoryImpl.java]: **Just an idea:** it is also convenient to use `@RequiredArgsConstructor` instead.

File: service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+        }
+    }
+
+    private GiftCertificate getGiftCertificateFromDto(GiftCertificateDto dto) {
+        GiftCertificate certificate = new GiftCertificate(dto.getId(), dto.getName(), dto.getDescription(), dto.getPrice(), dto.getDuration());
+        if (dto.getCreateDate() == null) {
+            certificate.setCreateDate(LocalDateTime.now());
+        } else {
+            certificate.setCreateDate(LocalDateTime.parse(dto.getCreateDate()));
+        }
```
Comment (maximelyashevich) [inline service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java]: What's about ternary operator?

**Review by maximelyashevich — APPROVED** [review]
Reviewed from my side, good job.

Regards,
Maks

File: repository/pom.xml (hunk 1 above)
Comment (letucennik) [reply repository/pom.xml]: removed it

File: repository/src/test/java/com/epam/esm/repository/impl/CertificateTagRepositoryImplTest.java (hunk 2 above)
Comment (letucennik) [reply repository/src/test/java/com/epam/esm/repository/impl/CertificateTagRepositoryImplTest.java]: done

File: repository/src/test/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImplTest.java (hunk 3 above)
Comment (letucennik) [reply repository/src/test/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImplTest.java]: done

File: repository/src/test/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImplTest.java (hunk 4 above)
Comment (letucennik) [reply repository/src/test/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImplTest.java]: done

File: repository/src/test/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImplTest.java (hunk 5 above)
Comment (letucennik) [reply repository/src/test/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImplTest.java]: done

File: repository/src/test/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImplTest.java (hunk 6 above)
Comment (letucennik) [reply repository/src/test/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImplTest.java]: done

File: repository/src/test/java/com/epam/esm/repository/impl/TagRepositoryImplTest.java (hunk 7 above)
Comment (letucennik) [reply repository/src/test/java/com/epam/esm/repository/impl/TagRepositoryImplTest.java]: done

File: repository/src/test/java/com/epam/esm/repository/impl/TagRepositoryImplTest.java (hunk 8 above)
Comment (letucennik) [reply repository/src/test/java/com/epam/esm/repository/impl/TagRepositoryImplTest.java]: done

File: service/pom.xml (hunk 9 above)
Comment (letucennik) [reply service/pom.xml]: no, removed it

File: web/pom.xml (hunk 10 above)
Comment (letucennik) [reply web/pom.xml]: only war is needed, removed the rest

File: repository/src/main/java/com/epam/esm/config/DataSourceConfig.java (hunk 11 above)
Comment (letucennik) [reply repository/src/main/java/com/epam/esm/config/DataSourceConfig.java]: Autowired annotation is not needed, removed it

File: service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java (hunk 13 above)
Comment (letucennik) [reply service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java]: good idea)

File: repository/src/main/java/com/epam/esm/repository/impl/CertificateTagRepositoryImpl.java (hunk 12 above)
Comment (letucennik) [reply repository/src/main/java/com/epam/esm/repository/impl/CertificateTagRepositoryImpl.java]: done

**Review by SenPavel — APPROVED** [review]

**Pull request merged**